src/sage/rings/polynomial/multi_polynomial_libsingular.pyx

```diff
@@ -3503,163 +3502,122 @@ cdef class MPolynomial_libsingular(MPolynomial):
+                    if  degree > _ring.bitmask:
                         id_Delete(&to_id, _ring)
                         p_Delete(&_p, _ring)
-                        raise OverflowError("exponent overflow (%d)"%(degree))
-                    need_map = 1
-
-                if _p == NULL:
-                    # the polynomial is 0
-                    try_symbolic = 1
-                    break
+                        raise OverflowError("exponent overflow (%d)" % (degree,))
+                    to_id.m[i] = p_Copy(_f, _ring)
 
-            if need_map:
-                for mi from 0 <= mi < _ring.N:
-                    if to_id.m[mi] == NULL:
-                        to_id.m[mi] = p_ISet(1,_ring)
-                        p_SetExp(to_id.m[mi], mi+1, 1, _ring)
-                        p_Setm(to_id.m[mi], _ring)
+        if not change_ring and need_map:
+            for mi from 0 <= mi < _ring.N:
+                if to_id.m[mi] == NULL:
+                    to_id.m[mi] = p_ISet(1,_ring)
+                    p_SetExp(to_id.m[mi], mi + 1, 1, _ring)
+                    p_Setm(to_id.m[mi], _ring)
 
-                from_id=idInit(1,1)
-                from_id.m[0] = _p
+            from_id = idInit(1, 1)
+            from_id.m[0] = _p
 
-                rChangeCurrRing(_ring)
-                res_id = fast_map_common_subexp(from_id, _ring, to_id, _ring)
-                _p = res_id.m[0]
+            rChangeCurrRing(_ring)
+            res_id = fast_map_common_subexp(from_id, _ring, to_id, _ring)
+            _p = res_id.m[0]
 
-                from_id.m[0] = NULL
-                res_id.m[0] = NULL
+            from_id.m[0] = NULL
+            res_id.m[0] = NULL
 
-                id_Delete(&from_id, _ring)
-                id_Delete(&res_id, _ring)
+            id_Delete(&from_id, _ring)
+            id_Delete(&res_id, _ring)
 
         id_Delete(&to_id, _ring)
 
-        if not try_symbolic:
+        if not change_ring:
             return new_MP(parent,_p)
 
-        # now as everything else failed, try to do it symbolically with call
-
-        cdef list g = list(parent.gens())
-
-        if fixed is not None:
-            for m,v in fixed.items():
-                if isinstance(m, (int, Integer)):
-                    mi = m+1
-                elif isinstance(m, MPolynomial_libsingular) and m.parent() is parent:
-                    for i from 0 < i <= _ring.N:
-                        if p_GetExp((<MPolynomial_libsingular>m)._poly, i, _ring) != 0:
-                            mi = i
-                            break
-                    if i > _ring.N:
-                        raise TypeError("key does not match")
-                else:
-                    raise TypeError("keys do not match self's parent")
+        # finally change the parent of the result
 
-                g[mi-1] = v
+        res_parent = coercion_model.common_parent(parent._base, *[v for _, v in items])
 
```

Comment:
This appears to be more general than the old code that tries to coerce every `v` into `parent._base`.
Is this just incidental, or is there a use case for this greater generality?

Reply:
Perhaps there is no difference in this respect. Both first try to coerce every `v` into `parent._base`. If fails, the old code finds the common parent in `return self(*g)` (the last line) via the constructor `__init__` while the new code does the same in the place. However, for some paths (substitutions with constant values), the new code does not make a duplicate call to `singular_polynomial_call` by avoiding going through the constructor. (if I remember correctly...)

Reply:
Perhaps it is the substitution with constant values that results in zero that makes the duplicate call.